File: doc-examples/example-java/src/main/java/example/MultipartUploadService.java
package example;

import io.micronaut.objectstorage.multipart.AbortMultipartUploadRequest;
import io.micronaut.objectstorage.multipart.CompleteMultipartUploadRequest;
import io.micronaut.objectstorage.multipart.CreateMultipartUploadRequest;
import io.micronaut.objectstorage.multipart.MultipartObjectStorageOperations;
import io.micronaut.objectstorage.multipart.UploadPartRequest;
import io.micronaut.objectstorage.request.UploadRequest;
import jakarta.inject.Singleton;

import java.util.List;

@Singleton
public class MultipartUploadService {

    private final MultipartObjectStorageOperations<?, ?, ?> multipartObjectStorage;

    public MultipartUploadService(MultipartObjectStorageOperations<?, ?, ?> multipartObjectStorage) {
        this.multipartObjectStorage = multipartObjectStorage;
    }

    //tag::multipart-upload[]
    public String uploadVideo(String key, byte[] nonFinalChunk, byte[] finalChunk) {
        var createResponse = multipartObjectStorage.createMultipartUpload(
            new CreateMultipartUploadRequest(key, "video/mp4")
        );
        var upload = createResponse.getUpload();

        try {
            var firstPart = multipartObjectStorage.uploadPart(
                new UploadPartRequest(upload, 1, UploadRequest.fromBytes(nonFinalChunk, upload.getKey()))
            );
            var secondPart = multipartObjectStorage.uploadPart(
                new UploadPartRequest(upload, 2, UploadRequest.fromBytes(finalChunk, upload.getKey()))
            );

            multipartObjectStorage.completeMultipartUpload(
                new CompleteMultipartUploadRequest(upload, List.of(firstPart.getPart(), secondPart.getPart()))
            );
            return upload.getKey();
        } catch (RuntimeException e) {
            multipartObjectStorage.abortMultipartUpload(new AbortMultipartUploadRequest(upload));
            throw e;
        }
    }
    //end::multipart-upload[]
}

File: doc-examples/example-python/src/main/python/example/MultipartUploadService.py
from jakarta.inject import Singleton
from java.lang import RuntimeException
from micronaut.objectstorage.multipart import (
    AbortMultipartUploadRequest,
    CompleteMultipartUploadRequest,
    CreateMultipartUploadRequest,
    MultipartObjectStorageOperations,
    UploadPartRequest,
)
from micronaut.objectstorage.request import UploadRequest


@Singleton
class MultipartUploadService:

    def __init__(self, multipart_object_storage: MultipartObjectStorageOperations):
        self.multipart_object_storage = multipart_object_storage

    # tag::multipart-upload[]
    def upload_video(self, key: str, non_final_chunk: bytes, final_chunk: bytes) -> str:
        create_response = self.multipart_object_storage.createMultipartUpload(
            CreateMultipartUploadRequest(key, "video/mp4")
        )
        upload = create_response.getUpload()

        try:
            first_part = self.multipart_object_storage.uploadPart(
                UploadPartRequest(upload, 1, UploadRequest.fromBytes(non_final_chunk, upload.getKey()))
            )
            second_part = self.multipart_object_storage.uploadPart(
                UploadPartRequest(upload, 2, UploadRequest.fromBytes(final_chunk, upload.getKey()))
            )

            self.multipart_object_storage.completeMultipartUpload(
                CompleteMultipartUploadRequest(upload, [first_part.getPart(), second_part.getPart()])
            )
            return upload.getKey()
        except RuntimeException as e:
            self.multipart_object_storage.abortMultipartUpload(AbortMultipartUploadRequest(upload))
            raise e
    # end::multipart-upload[]
